### ISSUE TITLE
《Neural Symbolic Machines》Discussion: What are the input and target for the neural net ?

### ISSUE THREAD

**Issue by guotong1988** [carpedm20/paper-notes#1]
https://github.com/carpedm20/paper-notes/blob/master/notes/neural-symbolic-machine.md

![image](https://user-images.githubusercontent.com/4702353/36824435-3ff86cbc-1d3d-11e8-9eb5-802c3a255387.png)

From the picture of paper, we find that argmax operation is in the target.
But the WEBQUESTIONSSP dataset is like this:
```
    {
      "QuestionId": "WebQTrn-2632",
      "RawQuestion": "what city has the largest population in mexico?",
      "ProcessedQuestion": "what city has the largest population in mexico",
      "Parses": [
        {
          "ParseId": "WebQTrn-2632.P0",
          "AnnotatorId": 2,
          "AnnotatorComment": {
            "ParseQuality": "Complete",
            "QuestionQuality": "Good",
            "Confidence": "Normal",
            "FreeFormComment": "can't do max after selecting \"city\""
          },
          "Sparql": "PREFIX ns: <http://rdf.freebase.com/ns/>\nSELECT DISTINCT ?x\nWHERE {\nFILTER (?x != ns:m.0b90_r)\nFILTER (!isLiteral(?x) OR lang(?x) = '' OR langMatches(lang(?x), 'en'))\nns:m.0b90_r ns:location.location.contains ?x .\n?x ns:common.topic.notable_types ns:m.01m9 .\n?x ns:topic_server.population_number ?sk0 .\n}\nORDER BY DESC(xsd:float(?sk0))\nLIMIT 1\n",
          "PotentialTopicEntityMention": "mexico",
          "TopicEntityName": "Mexico",
          "TopicEntityMid": "m.0b90_r",
          "InferentialChain": [
            "location.location.contains"
          ],
          "Constraints": [
            {
              "Operator": "Equal",
              "ArgumentType": "Entity",
              "Argument": "m.01m9",
              "EntityName": "City/Town/Village",
              "SourceNodeIndex": 0,
              "NodePredicate": "common.topic.notable_types",
              "ValueType": "String"
            }
          ],
          "Time": null,
          "Order": {
            "SortOrder": "Descending",
            "Start": 0,
            "Count": 1,
            "SourceNodeIndex": 0,
            "NodePredicate": "topic_server.population_number",
            "ValueType": "Number"
          },
          "Answers": [
            {
              "AnswerType": "Entity",
              "AnswerArgument": "m.04sqj",
              "EntityName": "Mexico City"
            }
          ]
        }
      ]
    },
```
@carpedm20 Thank you.

**Comment by guotong1988** [comment]
One of the author Joberant reply: 
Input is 'raw question' and output is 'answers'

**Comment by guotong1988** [comment]
ChenLiang reply:

The input of the neural network is the tokenized question (for example, ["what", "is", "the", "name", "of", "justin", "bieber", "brother", "?"]) and the entities resolved from the sentences by an entity linker (for example, the entity "justin bieber" m.06w2sn5). We use only the "Answers" (for example, the entity "Jaxon Bieber"  m.0gxnnwq) as weak supervision to train the model. The model is trained using RL. Each action is outputting a token and the reward is the F1 score of the generated program. There are more details about the RL training in section 2.3 of the paper. I also attached my ACL [slides](https://download.csdn.net/download/guotong1988/10281050).
